pandas/io/sql.py

```diff
@@ -729,41 +728,66 @@ def has_table(table_name: str, con, schema: str | None = None) -> bool:
     -------
     boolean
     """
-    pandas_sql = pandasSQL_builder(con, schema=schema)
-    return pandas_sql.has_table(table_name)
+    with pandasSQL_builder(con, schema=schema) as pandas_sql:
+        return pandas_sql.has_table(table_name)
 
 
 table_exists = has_table
 
 
-def pandasSQL_builder(con, schema: str | None = None) -> PandasSQL:
+@contextmanager
+def pandasSQL_builder(
+    con,
+    schema: str | None = None,
+    need_transaction: bool = False,
+) -> Iterator[PandasSQL]:
     """
     Convenience function to return the correct PandasSQL subclass based on the
-    provided parameters.
+    provided parameters.  Also creates a sqlalchemy connection and transaction
+    if necessary.
     """
     import sqlite3
 
     if isinstance(con, sqlite3.Connection) or con is None:
-        return SQLiteDatabase(con)
-
-    sqlalchemy = import_optional_dependency("sqlalchemy", errors="ignore")
+        yield SQLiteDatabase(con)
+    else:
+        sqlalchemy = import_optional_dependency("sqlalchemy", errors="ignore")
 
-    if isinstance(con, str):
-        if sqlalchemy is None:
+        if sqlalchemy is not None and isinstance(
+            con, (str, sqlalchemy.engine.Connectable)
+        ):
+            with _sqlalchemy_con(con, need_transaction) as con:
+                yield SQLDatabase(con, schema=schema)
+        elif isinstance(con, str) and sqlalchemy is None:
             raise ImportError("Using URI string without sqlalchemy installed.")
-        con = sqlalchemy.create_engine(con)
+        else:
 
-    if sqlalchemy is not None and isinstance(con, sqlalchemy.engine.Connectable):
-        return SQLDatabase(con, schema=schema)
+            warnings.warn(
+                "pandas only supports SQLAlchemy connectable (engine/connection) or "
+                "database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 "
+                "objects are not tested. Please consider using SQLAlchemy.",
+                UserWarning,
+                stacklevel=find_stack_level() + 2,
+            )
+            yield SQLiteDatabase(con)
+
+
+@contextmanager
+def _sqlalchemy_con(connectable, need_transaction: bool):
```

Comment:
Will this be reused in your followup PR? I have a slight preference of just folding this into `pandasSQL_builder` for now until there's a need to share this connection logic

Reply:
I only plan for `_sqlalchemy_con` to be called by `pandasSQL_builder` and not by anything else.  I split it out to reduce the amount of indentation and keep the functions small.  There is an additional try/finally block and some additional if/else blocks in lines 772-800 here: https://github.com/cdcadman/pandas/blob/sql_fixes/pandas/io/sql.py .  I think I can fold `_sqlalchemy_con` into `pandasSQL_builder` if you still prefer that.

Reply:
Okay gotcha. I think it'll be fine to leave as a separate function with more code incoming.

However, could you replace `import sqlalchemy` here with `sqlalchemy = import_optional_dependnecy(..., errors="raise")`? Just so that if someone else tries using this function in the future it's known that sqlalchemy is required.

Reply:
Ok, I made the change.